spring-integration-core/src/test/java/org/springframework/integration/handler/SuspendingMessageHandlerTests.java

```diff
@@ -0,0 +1,197 @@
+		Message<?> received = outputChannel.receive(10000);
+		assertNotNull(received);
+		Object[] result = (Object[]) received.getPayload();
+		assertEquals("foo", result[0]);
+		assertEquals("bar", ((Message<?>) result[1]).getPayload());
+		assertEquals(0, suspensions.size());
+	}
+
+	@Test
+	public void testLateReply() throws Exception {
+		final SuspendingMessageHandler handler = new SuspendingMessageHandler(0);
+		QueueChannel outputChannel = new QueueChannel();
+		handler.setOutputChannel(outputChannel);
+		handler.setBeanFactory(mock(BeanFactory.class));
+		handler.afterPropertiesSet();
+		Executors.newSingleThreadExecutor().execute(new Runnable() {
+
+			@Override
+			public void run() {
+				handler.handleMessage(MessageBuilder.withPayload("foo").setCorrelationId("foo").build());
+			}
+		});
+		Map<?, ?> suspensions = TestUtils.getPropertyValue(handler, "suspensions", Map.class);
+		int n = 0;
+		while (n++ < 100 && suspensions.size() == 0) {
+			Thread.sleep(100);
+		}
+		assertTrue("suspension did not appear in time", n < 100);
+		Log logger = spy(TestUtils.getPropertyValue(handler, "logger", Log.class));
+		new DirectFieldAccessor(handler).setPropertyValue("logger", logger);
+		handler.release(MessageBuilder.withPayload("bar").setCorrelationId("foo").build());
+		ArgumentCaptor<String> captor = ArgumentCaptor.forClass(String.class);
+		verify(logger).error(captor.capture());
+		assertThat(captor.getValue(),
+				Matchers.allOf(containsString("Release message arrived too late"),
+						containsString("payload=foo"), containsString("payload=bar")));
+		assertEquals(0, suspensions.size());
+	}
+
+	@Test
+	public void testExceptionReply() throws Exception {
+		final SuspendingMessageHandler handler = new SuspendingMessageHandler(10000);
+		QueueChannel outputChannel = new QueueChannel();
+		handler.setOutputChannel(outputChannel);
+		handler.setBeanFactory(mock(BeanFactory.class));
+		handler.afterPropertiesSet();
+		final AtomicReference<Exception> exception = new AtomicReference<Exception>();
+		final CountDownLatch latch = new CountDownLatch(1);
+		Executors.newSingleThreadExecutor().execute(new Runnable() {
+
+			@Override
+			public void run() {
+				try {
+					handler.handleMessage(MessageBuilder.withPayload("foo").setCorrelationId("foo").build());
+				}
+				catch (Exception e) {
+					exception.set(e);
+					latch.countDown();
+				}
+			}
+		});
+		Map<?, ?> suspensions = TestUtils.getPropertyValue(handler, "suspensions", Map.class);
+		int n = 0;
+		while (n++ < 100 && suspensions.size() == 0) {
+			Thread.sleep(100);
+		}
+		assertTrue("suspension did not appear in time", n < 100);
+		Exception exc = new RuntimeException();
+		handler.release(MessageBuilder.withPayload(exc).setCorrelationId("foo").build());
+		assertTrue(latch.await(10, TimeUnit.SECONDS));
+		assertSame(exc, exception.get().getCause());
+		assertEquals(0, suspensions.size());
+	}
+
```

Comment:
... and need the test to cover my use-case for the second `requestMessage` after `timeout`.